### DISCUSSION TITLE
86box as debugger

### DISCUSSION THREAD

**Discussion by bleuge** [86Box/86Box#1725] [Ideas]
The idea is exposing a simple debug API through as a server, so any external program could trace and request cpu info for whatever purpose, debugging, execution logging, getting stats, and other things.

I don't know too much about dynarecs, but read and studied many cpu different emulators, so I don't know if this could be easily enabled in non-dynarec modes. Just pausing the emulation, and listening to debug port and answering API requests. Basic functions could be step opcode, get cpu info, timings, exposing i/o ports and memory, etc.

Could this be done without too much code refactoring? Interesting for anyone?

The idea occured to me while playing with Bearlibterminal and thinking about coding a turbo debug/softice clone in a nice modern TUI.

**Comment by OBattler** [comment]
I'm not sure how feasible it is to step the code in a recompiler, the emulator somehow does it with the trap flag already, but I have no idea how.

**Comment by bleuge** [comment]
I suppose dealing with recompiler it will be hard. But I've reading the code and
Could I put logic in here to wait for commands via the tcpip server? exposing the debug api here? (forget about dynarec, just plain cores) 
Or timers would go crazy if the code sits here waiting?
```
                if (is386) {
#ifdef USE_DYNAREC
                        if (cpu_use_dynarec)
                                exec386_dynarec(clockrate/100);
                          else
#endif
                                exec386(clockrate/100);
                } else if (machines[machine].cpu[cpu_manufacturer].cpus[cpu_effective].cpu_type >= CPU_286) {
                        exec386(clockrate/100);
                } else {
                        execx86(clockrate/100);
                }
```
